main_test.py

```diff
@@ -18,54 +18,123 @@
 # These tests are unit tests that mock Pub/Sub.
 
 import base64
-from uuid import uuid4
+import re
 
 import pytest
 
 import main
+import philips_hue
+import philips_hue_mock
 
 
 @pytest.fixture
-def client():
+def config():
+    main.app.config.from_object('config.DevConfig')
+    return main.app.config
+
+
+@pytest.fixture
+def flask_client():
     main.app.testing = True
     return main.app.test_client()
 
 
 @pytest.fixture
-def config():
-    main.app.config.from_object('config.DevConfig')
-    return main.app.config
+def philips_hue_client(config):
+    philips_hue_client = philips_hue.PhilipsHueClient(
+        config['BRIDGE_IP_ADDRESS'], config['USERNAME'])
+    return philips_hue_client
+
+
+def test_empty_payload(flask_client):
+    response = flask_client.post('/', json='')
+
+    assert response.status_code == 400
+    assert b'invalid Pub/Sub message format' in response.data
+
+
+def test_invalid_payload(flask_client):
+    response = flask_client.post('/', json={'nomessage': 'invalid'})
+
+    assert response.status_code == 400
+    assert b'invalid Pub/Sub message format' in response.data
+
+
+def test_invalid_mimetype(flask_client):
+    response = flask_client.post('/', json="{ message: true }")
+
+    assert response.status_code == 400
+    assert b'invalid Pub/Sub message format' in response.data
+
+
+def test_nonstring_pubsub_message(flask_client):
+    response = flask_client.post('/', json={'message': True})
+
+    assert response.status_code == 400
+    assert b'invalid Pub/Sub message format' in response.data
+
+
+def test_nonstring_notification_message(flask_client):
+    response = flask_client.post('/', json={'message': {'data': True}})
+
+    assert response.status_code == 400
+    assert b'data should be in a string format' in response.data
+
+
+def test_unicode_notification_message(flask_client):
+    data = '{"incident": {"condition": {"state": "open"}}}'
```

Comment:
What if we include a character outside of the normally supported ASCII values? See http://kellykjones.tripod.com/webtools/ascii_utf8_table.html. Notice the nearly identical encoding columns.

Reply:
done